Validators list filter search

eshark9312 opened the pull request:
Resolves #4328 
Resolves #4329 
Add the components for validators list and filter including search box.

thesan commented:
@eshark9312 same as https://github.com/Joystream/pioneer/pull/4418#issuecomment-1589373761 because this PR is based on your #4375, I'm converting it to draft for the moment.

thesan reviewed (changes requested):
@eshark9312 please add some verified and waiting validator in the story.

The goal is to add some automated tests to check:
1. If the validator are showing how it should be showing
2. If the filter works

Also please add a test for the search and clear filter.

I think all of these can be in one test flow but please add it in a different story (but the same file). So in `packages/ui/src/app/pages/Validators/ValidatorList.stories.tsx` add a:
```ts
export const TestsFilters: Story = {
  play: async ({ canvasElement, step }) => {
    // Test code...
  }
}

eshark9312 commented:
- I fixed the [storybook](https://pioneer-2-storybook-9b0x8grbw-joystream.vercel.app/?path=/story/pages-validators-validatorlist--statistics-and-lists) so there are some waiting validators in the list now.
Also I add some verified validators by modified the custom hook `useValidatorsList.tsx`.
Since the validator's state (`verified/unverified`) can't be fetched from the current chain nor qn, as you know, I added verification state for each validator randomly: `verification: Math.random() > 0.5`

- I added a [test](https://pioneer-2-storybook-9b0x8grbw-joystream.vercel.app/?path=/story/pages-validators-validatorlist--tests-filters) for the search and filter.
But there is a small problem which seems to be due to the version of the packages to me, not sure.
`userEvent.type(searchElement, '{enter}')` does not work, so no changes from search result in the automated test.

thesan commented on packages/ui/src/app/pages/Validators/ValidatorList.stories.tsx:
```diff
@@ -113,4 +142,44 @@ export default {
+    const searchElement = screen.getByPlaceholderText('Search')
+    const verificationFilter = screen.getAllByText('Verification')[0]
+    const stateFilter = screen.getAllByText('State')[0]
+    const clearFilter = () => {
+      userEvent.click(screen.getByText('Clear all filters'))
+    }
```
This should be awaited even though `userEvent.click` doesn't return a promise (my guess is that it's because storybook monkeypatches these methods when they're called in the storybook ui).

Also IMO it would be better to save `screen.getByText('Clear all filters')` in a constant (just like how you did it for `searchElement`, `verificationFilter`, and `stateFilter`. And then call  `await userEvent.click(clearFilter)` every time the tests need to clear the filters.

eshark9312 replied:
I fixed the code according to your review. `screen.getByText('Clear all filters')` occurs error when I define it as a constant, because it does not appear when all validators are shown.

thesan reviewed (changes requested):
In [the design](https://www.figma.com/file/tjad3r2EZJXRMFOELcxxNR/Validators?type=design&node-id=146-94799):
![image](https://github.com/Joystream/pioneer/assets/6571453/723c6c25-61bc-4507-a4f1-c60522b943ae)

Each verified validator has a name + extra info. I think there should have been another issue before #4404 which should have been just about displaying the validator list (without the filters) but something went wrong and we forgot to create this issue. As a result it will have to be done in this PR. Meaning Please add the necessary queries to get the information about verified validators.

thesan commented:
@eshark9312 btw one of the reason adding the verified validator data is needed here is because the search needs to apply to these data too. E.g users might not know the account addresses but most of the time they will try to search the validators handle instead

eshark9312 commented on packages/ui/src/app/pages/Validators/ValidatorList.stories.tsx:
```diff
@@ -113,4 +142,44 @@ export default {
+      userEvent.click(screen.getByText('Clear all filters'))
+    }
+
+    await step('Verifcation Filter', async () => {
+      await selectFromDropdown(screen, verificationFilter, 'verified')
+      await selectFromDropdown(screen, verificationFilter, 'unverified')
+      await selectFromDropdown(screen, verificationFilter, 'All')
+    })
+    await step('State Filter', async () => {
+      await selectFromDropdown(screen, stateFilter, 'active')
+      await selectFromDropdown(screen, stateFilter, 'waiting')
+      await selectFromDropdown(screen, stateFilter, 'All')
+    })
+    await step('Search', async () => {
+      userEvent.type(searchElement, 'j4RLnWh{enter}')
+      userEvent.type(searchElement, 'j4Rx{enter}')
+    })
+
+    await step('Clear Filter', async () => {
+      await selectFromDropdown(screen, verificationFilter, 'verified')
+      await selectFromDropdown(screen, stateFilter, 'waiting')
+      clearFilter()
+      await selectFromDropdown(screen, stateFilter, 'active')
+      clearFilter()
+      await selectFromDropdown(screen, verificationFilter, 'unverified')
+      await selectFromDropdown(screen, stateFilter, 'waiting')
+      userEvent.type(searchElement, 'j4Rx{enter}')
+      clearFilter()
+    })
```
I added some code to verify the filtering results.

eshark9312 commented on packages/ui/src/validators/hooks/useValidatorsList.tsx:
```diff
@@ -0,0 +1,94 @@
+        return {
+          member: false,
+          address: encodeAddress(address),
+          verification: Math.random() > 0.5,
```
As the validators have no membership now, so there is no information about validators in the query node yet.

eshark9312 replied:
I think it needs to extend the query node  to include the information(`isValidator, verification, startedon`) about validators.

thesan commented on packages/ui/src/app/pages/Validators/ValidatorList.stories.tsx:
```diff
@@ -113,4 +143,49 @@ export default {
+      await selectFromDropdown(screen, verificationFilter, 'All')
+    })
+    await step('State Filter', async () => {
+      await selectFromDropdown(screen, stateFilter, 'active')
+      await expect(screen.queryByText('waiting')).toBeNull()
+      await selectFromDropdown(screen, stateFilter, 'waiting')
+      await expect(screen.queryByText('active')).toBeNull()
+      await selectFromDropdown(screen, stateFilter, 'All')
+    })
+    await step('Search', async () => {
+      await userEvent.type(searchElement, 'j4RLnWh{enter}')
+      await userEvent.type(searchElement, 'j4Rx{enter}')
+    })
+
+    await step('Clear Filter', async () => {
+      await selectFromDropdown(screen, verificationFilter, 'verified')
+      await selectFromDropdown(screen, stateFilter, 'waiting')
+      await clearFilter()
+      await selectFromDropdown(screen, stateFilter, 'active')
+      await clearFilter()
+      await selectFromDropdown(screen, verificationFilter, 'unverified')
+      await selectFromDropdown(screen, stateFilter, 'waiting')
+      await userEvent.type(searchElement, 'j4Rx{enter}')
+      await clearFilter()
```
Same thing for the other tests please

eshark9312 replied:
ok, I fixed

eshark9312 commented on packages/ui/src/validators/components/ValidatorItem.tsx:
```diff
@@ -0,0 +1,71 @@
+        </BadgeStatus>
+        <TokenValue size="xs" value={totalRewards} />
+        <TextMedium bold> {APR} </TextMedium>
+        <TextSmall> Dec 2022 </TextSmall>
```
Sorry, I forgot about this problem.
Unfortunately, there is no other way to determine the exact time when a validator started working without exploring all blocks in the blockchain. Same as [`totalReward`](https://github.com/Joystream/pioneer/issues/4326#issuecomment-1560402793)
I guess it should be added in query-node with validator membership.

eshark9312 commented:
@thesan 
There is a problem during the storybook test, I guess it is because of the version of the packages, as I mentioned before.
userEvent.type(searchElement, '{enter}') does not work, so the search results are not displayed in the automated test.

thesan commented on packages/ui/src/validators/hooks/useValidatorsList.tsx:
```diff
@@ -0,0 +1,161 @@
+                  .mul(validatorInfo.commission.toBn())
+                  .div(new BN('1000000000'))
+                  .div(stakingInfo.total.toBn())
+                  .toNumber(),
```
Thanks a lot for clarifying that part I was completely lost !

So the code is easier to understand in the future, here's the alternative I'm proposing:
```ts
const apr =
  rewardHistory.length &&
  stakingInfo.total.toNumber() &&
  last(rewardHistory).eraReward.toBn()
    .muln(ERAS_PER_YEAR)
    .mul(validatorInfo.commission.toBn())
    .div(stakingInfo.total.toBn())
    .divn(10 ** 7) // Convert from Perbill to Percent
    .toNumber()
```
Please define `ERAS_PER_YEAR` in `@/common/constants/numbers.ts`

eshark9312 replied:
I fixed the code according to your proposal

thesan commented:
> @thesan There is a problem during the storybook test, I guess it is because of the version of the packages, as I mentioned before. userEvent.type(searchElement, '{enter}') does not work, so the search results are not displayed in the automated test.

It turned out to be due to the search value (and validation) being debounced. I fixed by it pressing entered until it works !

thesan commented on packages/ui/src/app/pages/Validators/ValidatorList.stories.tsx:
```diff
@@ -113,4 +143,78 @@ export default {
+    const expectNumberOfValidatorsToBe = async (n: number) => {
+      const liElements = screen.queryAllByRole('listitem')
+      const validatorElements = liElements.filter((li) => li.textContent?.includes('Nominate'))
+      await expect(validatorElements.length).toBe(n)
```
No need for the `await` here so no need for the function's `async` either (and the `await`s when calling it).

But also `expect(screen.queryAllByRole('button', { name: 'Nominate' })).toHaveLength(n)` instead of `expectNumberOfValidatorsToBe(n)` is just as good. In tests I think the code should be as basic as possible (as long as it doesn't create too much boilerplate)

thesan commented on packages/ui/src/app/pages/Validators/ValidatorList.stories.tsx:
```diff
@@ -113,4 +143,78 @@ export default {
+    const searchElement = screen.getByPlaceholderText('Search')
+    const verificationFilter = screen.getAllByText('Verification')[0]
+    const stateFilter = screen.getAllByText('State')[0]
+    const clearFilter = async () => {
+      expect(screen.getByText('Clear all filters'))
+      await userEvent.click(screen.getByText('Clear all filters'))
+    }
```
Same here `getByText` is synchronous so no need for the `await` but IMO calling `userEvent.click(screen.getByText('Clear all filters'))` everywhere instead of `clearFilter()` is better

thesan commented on packages/ui/src/app/pages/Validators/ValidatorList.stories.tsx:
```diff
@@ -113,4 +143,78 @@ export default {
+
+    await step('Verifcation Filter', async () => {
+      await selectFromDropdown(screen, verificationFilter, 'verified')
+      await expect(screen.queryByText('unverifed')).toBeNull()
```
The `queryBy...` methods are synchronous too. Only `findBy...` are async.
```suggestion
      expect(screen.queryByText('unverifed')).toBeNull()
```

thesan commented on packages/ui/src/app/pages/Validators/ValidatorList.stories.tsx:
```diff
@@ -113,4 +143,78 @@ export default {
+      await selectFromDropdown(screen, verificationFilter, 'verified')
+      await expect(screen.queryByText('unverifed')).toBeNull()
+      await expectNumberOfValidatorsToBe(4)
+      await expect(screen.getByText('alice'))
+      await expect(screen.getByText('bob'))
```
```suggestion
      expect(screen.getByText('alice'))
      expect(screen.getByText('bob'))
```

Please remove the `await` in all the screen `getBy...` and `queryBy...` call after these ones.

thesan commented on packages/ui/src/app/pages/Validators/ValidatorList.stories.tsx:
```diff
@@ -113,4 +143,78 @@ export default {
+      await expectNumberOfValidatorsToBe(4)
+      await clearFilter()
+      await expectNumberOfValidatorsToBe(11)
+      await userEvent.type(searchElement, 'j4R{enter}')
+      await expectNumberOfValidatorsToBe(9)
```
I have no idea why this one appear to work but to be safe I think it should be changed too:
```suggestion
      await userEvent.type(searchElement, 'j4R')
      await waitFor(async () => {
        await userEvent.type(searchElement, '{enter}')
        expect(screen.queryAllByRole('button', { name: 'Nominate' })).toHaveLength(9)
      })
```

eshark9312 replied:
thanks

thesan reviewed (changes requested):
The test logic is good. There is just some small syntax issues. When `eslint` is setup correctly it should notice the unnecessary `await`s but still enforce the `await userEvent...` (thanks to the eslint plugin `storybook/recommended`). It works in my IDE but for some reason it's ignored by the CI and the commit hooks.

thesan commented:
Also on the APR: https://github.com/Joystream/pioneer/pull/4404#discussion_r1284641560

thesan commented:
I'll take a deeper look at the ["started on" problem](https://github.com/Joystream/pioneer/pull/4404#discussion_r1284537218) next week.

eshark9312 commented on packages/ui/src/app/pages/Validators/ValidatorList.stories.tsx:
```diff
@@ -99,10 +113,26 @@ export default {
+                  ['j4ShWRXxTG4K5Q5H7KXmdWN8HnaaLwppqM7GdiSwAy3eTLsJt'],
+                  ['j4WfB3TD4tFgrJpCmUi8P3wPp3EocyC5At9ZM2YUpmKGJ1FWM'],
+                ],
+                commission: joy(0.0000011),
```
thanks for your careful review, I fixed it

thesan commented on packages/ui/src/validators/hooks/useValidatorsList.tsx:
```diff
@@ -0,0 +1,165 @@
+            .muln(ERAS_PER_YEAR)
+            .mul(validatorInfo.commission.toBn())
+            .div(stakingInfo.total.toBn())
+            .divn(10 ** 7)
```
Please keep this comment because it's not obvious for me:
```suggestion
            .divn(10 ** 7) // Convert from Perbill to Percent
```

eshark9312 replied:
ok

thesan commented on packages/ui/src/validators/types/Validator.ts:
```diff
@@ -0,0 +1,14 @@
+  isActive: boolean
+  totalRewards: BN
+  APR: number
+  startedOn: string
```
I confirmed with @dmtrjsg that we can just remove the "started on" column.
```suggestion
```

eshark9312 replied:
I got it

thesan reviewed (approved):
Good job!
